Sources/Observer.swift

```diff
@@ -37,28 +25,46 @@ extension Signal {
 		/// - parameters:
 		///   - observer: The observer to transform.
 		///   - transform: The transform.
-		///   - disposable: The disposable to be disposed of when the `TransformerCore`
-		///                 yields any terminal event. If `observer` is a `Signal` input
-		///                 observer, this can be omitted.
+		///   - disposables: The disposable to be disposed of upon termination.
 		internal init<U, E>(
-			_ observer: Signal<U, E>.Observer,
-			_ transform: @escaping Event.Transformation<U, E>,
-			_ disposable: Disposable? = nil
+			producerObserver: Signal<U, E>.Observer,
+			applying transform: @escaping Event.Transformation<U, E>,
+			disposables: CompositeDisposable
 		) {
 			var hasDeliveredTerminalEvent = false
 
-			self._send = transform { event in
+			let wrappedOutputSink: Signal<U, E>.Observer.Action = { event in
 				if !hasDeliveredTerminalEvent {
-					observer._send(event)
+					producerObserver._send(event)
 
 					if event.isTerminating {
 						hasDeliveredTerminalEvent = true
-						disposable?.dispose()
+						disposables.dispose()
 					}
 				}
 			}
 
-			self.wrapped = observer.interruptsOnDeinit ? observer : nil
+			self._send = transform(wrappedOutputSink, Lifetime(disposables))
+			self.interruptsOnDeinit = false
+		}
+
+		/// An initializer that transforms the action of the given observer with the
+		/// given transform.
+		///
+		/// If the given observer would perform side effect on deinitialization, the
+		/// created observer would retain it.
+		///
+		/// - parameters:
+		///   - observer: The observer to transform.
+		///   - transform: The transform.
+		///   - disposables: The disposable to be disposed of upon termination.
+		///                 Used by `SignalProducer` only, since `Signal` takes.
```

Comment:
This `init` doesn't have a `disposables` parameter. The line also seems incomplete.